src/expressions/method-call-expr.md

```diff
@@ -15,22 +15,158 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
-Then, for each candidate `T`, add `&T` and `&mut T` to the list immediately after `T`.
+The following procedure is used:
 
-For instance, if the receiver has type `Box<[i32;2]>`, then the candidate types will be `Box<[i32;2]>`, `&Box<[i32;2]>`, `&mut Box<[i32;2]>`, `[i32; 2]` (by dereferencing), `&[i32; 2]`, `&mut [i32; 2]`, `[i32]` (by unsized coercion), `&[i32]`, and finally `&mut [i32]`.
+## Determining candidate types
+
+First, a list of "candidate types" is assembled.
+
+These types are found by taking the receiver type and iterating, following either:
+
+* The built-in [dereference]; or
+* `<T as Receiver>::Target`
+
+to the next type. (If a step involved following the `Receiver` target, we also
+note whether it would have been reachable by following `<T as
+Deref::Target>` - this information is used later).
+
+At the end, an additional candidate step may be added for
+an [unsized coercion].
+
+Each step of this chain provides a possible `self` type for methods that
+might be called. The list will be used in two different ways:
+
+* To find types that might have methods. This is used in the
+  "determining candidate methods" step, described below. This considers
+  the full list.
+* To find types to which the receiver can be converted. This is used in the
+  "picking a method from the candidates" step, also described below - in this
+  case, we only consider the types reachable via `Deref` or built-in
+  dereferencing.
+
+There is a built-in implementation of `Receiver` for all `T: Deref`, so
+most of the time, every step can be reached through either mechanism.
+Sometimes, more types can be reached via the `Receiver` chain, and so
+more types will be considered for the former usage than the latter usage.
+
+For instance, if the receiver has type `Box<[i32;2]>`, then the candidate types
+will be `Box<[i32;2]>`,`[i32; 2]` (by dereferencing), and `[i32]` (by unsized
+coercion).
+
+If `SmartPtr<T>: Receiver<Target=T>`, and the receiver type is `&SmartPtr<Foo>`,
+then the candidate types would be `&SmartPtr<Foo>`, `SmartPtr<Foo>` and `Foo`.
+
+## Determining candidate methods
+
+This list of candidate types is then converted to a list of candidate methods.
+For each step, the candidate type is used to determine what searches to perform:
+
+* For a trait object, there is first a search for inherent candidates for
+  the object, then inherent impl candidates for the type.
+* For a struct, enum, or foreign type, there is a search for inherent
+  impl candidates for the type.
+* For a type param, there's a search for inherent candidates on the param.
+* For various simpler types (listed below) there's a search for inherent
+  candidates for the incoherent type.
+* After any of these, there's a further search for extension candidates for
+  traits in scope.
```

Comment:
As I understand it, the "further search for extension candidates for traits in scope." happens once, not once for each candidate type, so I suggest putting this after the bulleted list rather than as part of it.

Perhaps mention the special case described in https://github.com/rust-lang/reference/issues/630#issuecomment-1229568011 here?

Reply:
> the "further search for extension candidates for traits in scope." happens once, not once for each candidate type

Yep, it was my intent to say that, thanks for spotting it wasn't clear, I moved it.

> Perhaps mention the special case

Reading the code, I don't think this is a special case. I think that these are regarded as inherent methods. I think this relates to someone else's comment that the meaning of "inherent" and "extension" is unclear - I'll address that separately.

Reply:
I tried an [experiment](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5c3a53532416abac6b15e6f7ccd0e769) for the "a trait is considered for method resolution inside an impl of that trait" rule.

I think that's telling us that the methods which become visible via that rule are treated as extension methods.

Reply:
As far as I can tell the rule is simply that when we're inside a trait impl the trait the impl's for is "in scope" in the same way as a trait that's visible by being defined or `use`d in the current module would be.

I think this is being done by [`traits_in_scope()`](https://github.com/rust-lang/rust/blob/9cd60bd2ccc41bc898d2ad86728f14035d2df72d/compiler/rustc_resolve/src/lib.rs#L1755C1-L1761C1) in the name resolution code.